### PR TITLE
Logger Additional Features

### PR DESCRIPTION
**Pull request by Neche-Stephen**
- Converted the classes in src/log-entry.ts to interfaces and utilized the fields in flow.ts and chatgpt-log.ts.
- Created a new interface in log-entry.ts for BufferEntry.
- Updated processMeta to use camelcase in the field values for Meta data - over snake_case
- Updated the flushLogs() function to take an optional transport parameter. 
- Implemented transport for file and console in transports.ts file
- Moved the jsonl files to a data directory
- Changed LogEntryType in const.ts to an enum

### CODE REVIEW

File: src/const.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
-export const LogEntryType = Object.freeze({
-  PROMPT: "prompt",
-  REQUEST: "request",
-  RESPONSE: "response",
-  FUNCTION_CALL: "functionCall",
-  CUSTOM: "custom",
-  ERROR: "error",
-});
+export enum LogEntryType {
+  PROMPT = "prompt",
```
Comment (justiceo) [inline src/const.ts]: Remove the string assignments to simplify this enum to PROMPT, RESPONSE...

File: src/log-entry.ts [hunk 2]
```diff
@@ -134,12 +135,12 @@ export class Meta {
 /**
  * Represents a complete log entry.
  */
-export class LogEntry {
+export interface LogEntry {
   /** The ID of the request */
-  requestId?: string;
+  requestId?: string | null;
```
Comment (justiceo) [inline src/log-entry.ts]: The null is unnecessary

File: src/flow.ts [hunk 3]
```diff
@@ -68,34 +51,34 @@ class Flow {
       type: LogEntryType.PROMPT,
       timestamp: new Date().toISOString(),
       data: { prompt, trigger },
-    });
+    } );
   }
 
-  logRequest(requestData: RequestData): void {
+  logRequest(requestData: Request): void {
```
Comment (justiceo) [inline src/flow.ts]: Type should be "any". The Request object is created during logs processing

File: src/log-entry.ts [hunk 4]
```diff
@@ -153,3 +154,17 @@ export class LogEntry {
...
+ */
+export interface BufferEntry {
+  /** The type of the entry (e.g., "request", "response", "functionCall") */
+  type: string;
+
+  /** The timestamp of when the entry was created */
+  timestamp: string;
+
+  /** The actual data of the entry, which can be a request, response, function call, or any other type */
+  data: Request | Response | FunctionCall | any;
```
Comment (justiceo) [inline src/log-entry.ts]: Type should be any

File: src/chatgpt-log.ts [hunk 5]
```diff
@@ -78,17 +50,22 @@ export class ChatGptLog {
...
+  processMeta(buffer: Readonly<BufferEntry[]>): Meta {
     const metaDetails = buffer.find((entry) => entry.type === LogEntryType.CUSTOM);    
     return {
-      output_mode: metaDetails?.data?.data?.output_mode,
-      user_id: metaDetails?.data?.data?.user_id, 
-      country: metaDetails?.data?.data?.country,  
-      operating_system: metaDetails?.data?.data?.operating_system, 
-      shell: metaDetails?.data?.data?.shell, 
-      user_time_zone: metaDetails?.data?.data?.user_time_zone
+      totalTokenCount: metaDetails?.data?.data?.totalTokenCount,
```
Comment (justiceo) [inline src/chatgpt-log.ts:62]: This looks wrong, I don't think the chatgpt responses are structured like below

**Review by justiceo — CHANGES_REQUESTED** [review]
PTAL

File: src/log-entry.ts (hunk 2 above)
Comment (Neche-Stephen) [reply src/log-entry.ts]: this.sessionId and this.currentRequestId  are set to null in the flow class before being subsequently updated to string, Typescript will not accept that, if there types are not set to string | (or) null

File: src/chatgpt-log.ts (hunk 5 above)
Comment (Neche-Stephen) [reply src/chatgpt-log.ts:62]: - Most of the responses are not from chatgpt. It is created during the process of logging
- Then in logging meta or custom data, there is a data field - data: { key, data }, this adds an extra layer of "data" to access

File: src/flow.ts [hunk 6]
```diff
@@ -68,34 +51,34 @@ class Flow {
...
+  logRequest(requestData: any): void {
     this.buffer.push({
       type: LogEntryType.REQUEST,
       timestamp: new Date().toISOString(),
       data: requestData,
     });
   }
 
-  logResponse(responseData: any): void {
+  logResponse(responseData: Response): void {
```
Comment (justiceo) [inline src/flow.ts]: any type

File: src/flow.ts [hunk 7]
```diff
@@ -68,34 +51,34 @@ class Flow {
...
     this.buffer.push({
       type: LogEntryType.RESPONSE,
       timestamp: new Date().toISOString(),
       data: responseData,
     });
   }
 
-  logFunctionCall(functionCallData: FunctionCallData): void {
+  logFunctionCall(functionCallData: FunctionCall): void {
```
Comment (justiceo) [inline src/flow.ts]: any type

File: src/flow.ts [hunk 8]
```diff
@@ -68,34 +51,34 @@ class Flow {
...
     this.buffer.push({
       type: LogEntryType.FUNCTION_CALL,
       timestamp: new Date().toISOString(),
       data: functionCallData,
     });
   }
 
-  log(key: string, data: any): void {
+  log(key: string, data: Meta ): void {
```
Comment (justiceo) [inline src/flow.ts]: any

File: src/log-entry.ts (hunk 2 above)
Comment (justiceo) [reply src/log-entry.ts]: Set it to undefined instead of null.

File: src/chatgpt-log.ts (hunk 5 above)
Comment (justiceo) [reply src/chatgpt-log.ts:62]: This class should process the responses from chatgpt, if using mocks, you may have to update the mock responses to match a real chatgpt responses.

This also applies to the processRequest, response, functionCall methods.

**Review by justiceo — CHANGES_REQUESTED** [review]
Left some comments

File: src/chatgpt-log.ts [hunk 9]
```diff
@@ -46,49 +19,61 @@ export class ChatGptLog {
       topP: request?.data?.top_p,
       functionCalls: request?.data?.functions,
       maxTokens: request?.data?.max_tokens,
-      startTime: prompt?.timestamp,
-      sentTime: request?.timestamp,
+      startTime: request?.data.startTime,
```
Comment (justiceo) [inline src/chatgpt-log.ts]: I don't think that startTime and sentTime are in the request object. 

I originally had startTime as the time the prompt object was created => prompt.timestamp and sentTime as the time the request object was created => request.timestamp.

File: src/flow.ts [hunk 10]
```diff
@@ -36,8 +19,8 @@ class Flow {
   constructor() {
     // Initialize handlers in the constructor
     this.buffer = [];
-    this.sessionId = null;
-    this.currentRequestId = null;
+    this.sessionId = undefined;
+    this.currentRequestId = undefined;
```
Comment (justiceo) [inline src/flow.ts]: These are already undefined, no need to set them as undefined.

File: src/index.ts [hunk 11]
```diff
@@ -33,54 +23,59 @@ async function getWeather(location: string): Promise<WeatherData> {
   return weatherData;
 }
 
-async function useChatGptApi(prompt: string): Promise<void> {
+export async function useChatGptApi(prompt: string) {
```
Comment (justiceo) [inline src/index.ts:26]: Still planning to move this to the test file?

**Review by justiceo — CHANGES_REQUESTED** [review]
PTAL

File: src/index.ts (hunk 11 above)
Comment (Neche-Stephen) [reply src/index.ts:26]: I imported the method to the test file. So it has been moved. I did not to push the whole file, just the method that triggers the use of the logger and returns the logentry

File: src/chatgpt-log.ts (hunk 9 above)
Comment (Neche-Stephen) [reply src/chatgpt-log.ts]: I checked now, Start time is in the interface for Request, do want me to remove it?

File: src/flow.ts (hunk 10 above)
Comment (Neche-Stephen) [reply src/flow.ts]: Alright. Will remove now

File: src/index.ts (hunk 11 above)
Comment (justiceo) [reply src/index.ts:26]: I meant to move the entire contents of this file to spec (and delete this file).

**Review by justiceo — CHANGES_REQUESTED** [review]

**Pull request merged**